eth2/utils/bls/src/macros.rs

```diff
@@ -84,3 +84,112 @@ macro_rules! impl_cached_tree_hash {
         }
     };
 }
+
+macro_rules! bytes_struct {
+    ($name: ident, $type: ty, $byte_size: expr, $small_name: expr, $ssz_type_size: ident,
+     $type_str: expr, $byte_size_str: expr) => {
+        #[doc = "Stores `"]
+        #[doc = $byte_size_str]
+        #[doc = "` bytes which may or may not represent a valid BLS "]
+        #[doc = $small_name]
+        #[doc = ".\n\nThe `"]
+        #[doc = $type_str]
+        #[doc = "` struct performs validation when it is instantiated, where as this struct does \
+                 not. This struct is suitable where we may wish to store bytes that are \
+                 potentially not a valid "]
+        #[doc = $small_name]
+        #[doc = " (e.g., from the deposit contract)."]
+        #[derive(Clone)]
+        pub struct $name([u8; $byte_size]);
+    };
+    ($name: ident, $type: ty, $byte_size: expr, $small_name: expr, $ssz_type_size: ident) => {
+        bytes_struct!($name, $type, $byte_size, $small_name, $ssz_type_size, stringify!($type),
+        stringify!($byte_size));
+
+        impl $name {
+            pub fn from_bytes(bytes: &[u8]) -> Result<Self, ssz::DecodeError> {
+                Ok(Self(Self::get_bytes(bytes)?))
+            }
+
+            pub fn empty() -> Self {
+                Self([0; $byte_size])
+            }
+
+            pub fn as_bytes(&self) -> Vec<u8> {
+                self.0.to_vec()
+            }
+
+            fn get_bytes(bytes: &[u8]) -> Result<[u8; $byte_size], ssz::DecodeError> {
+                let mut result = [0; $byte_size];
+                if bytes.len() != $byte_size {
+                    Err(ssz::DecodeError::InvalidByteLength {
+                        len: bytes.len(),
+                        expected: $byte_size,
+                    })
+                } else {
+                    result[..].copy_from_slice(bytes);
+                    Ok(result)
+                }
+            }
+        }
+
+        impl std::fmt::Debug for $name {
+            fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                self.0[..].fmt(formatter)
+            }
+        }
+
+        impl PartialEq for $name {
+            fn eq(&self, other: &Self) -> bool {
+                &self.0[..] == &other.0[..]
+            }
+        }
+
+        impl Eq for $name {}
+
+        impl std::convert::TryInto<$type> for &$name {
+            type Error = ssz::DecodeError;
+
+            fn try_into(self) -> Result<$type, Self::Error> {
+                <$type>::from_bytes(&self.0[..])
+            }
+        }
+
+        impl std::convert::From<$type> for $name {
+            fn from(obj: $type) -> Self {
+                // how to avoid this unwrap? We know that obj.as_bytes() always has exactly
+                // $byte_size many bytes.
```

Comment:
I think an `unwrap` (or an `expect`) is OK here, as we know that the SSZ representation of a pubkey/signature is the right size, but can't encode that in a way the compiler will understand. If you add a test that checks `PublicKeyBytes::from(pubkey)` and `SignatureBytes::from(signature)` then we'll be sure to catch this issue if the SSZ size of these structs ever changes.

Reply:
I added now tests to check the length of a public key and a signature (but not in the ..Bytes file but directly in Signature and PublicKey). There are already tests for PublicKeyBytes::from and SignatureBytes::from (see PublicKeyBytes::tests::test_valid_public_key and SignatureBytes::tests::test_valid_signature)